doc/design/optimizer.md

```diff
@@ -0,0 +1,88 @@
+## Optimizer Design
+Optimizer is used to update parameter according to parameter gradient. 
+
+#### A typical training process is:
+
+1. run forward to calculate activation using data and parameter.
+1. run backward to calculate the gradient of activation and parameter using cost, activation, and parameter.
+1. apply the gradient to the corresponding parameter.
+
+User write code to describe the forward process, Optimizer should help implicitly add the process of backward and update according to forward process and other configuration.
+
+In PaddlePaddle, we use graph of operators to describe computation. So the Python class `Optimizer` defines the API to add Ops to computation block. It adds two kinds of Operators:
+
+1. Gradient Ops. Used to calculate the gradients for parameters.
+2. Optimize Ops. Used to apply gradient to parameters.
+
+The python user interface is:
+
+```python
+class Optimizer(object):
+	def _backward(loss):
+		"""
+		Add Operators to Compute gradients of `loss` 
+		It returns the backward net.
+		"""
+		...
+		return backward_net
+
+	def _update(var_list):
+		"""
+		Add Operators to Apply gradients to variables 
+		in var_list. It returns an update `Operator`.
+		"""
+		...
+		return update_op
+
+	def minimize(loss, var_list):
+		"""Add operations to minimize `loss` by updating `var_list`.
+		
+		This method simply combines calls `_backward()` and
+		`_update()`.
+		"""
+		backward_net = _backward(loss)
```

Comment:
What's the type of backward_net? NetOp, Block or var_list?
And does class optimizer has some private member? Is backward_net the member of Optimizer?

Reply:
This part has some problems, I will update it.